### PR TITLE
Add benchmarks for interaction fingerprints

### PR DESCRIPTION
**Pull request by ncfrey**
This PR adds scripts and benchmarks for predicting binding affinities from PDBbind using interaction fingerprints (via RdkitGridFeaturizer) and random forests / fully-connected networks.

Notes on the benchmark:
* The benchmarking runs are pretty slow because of the large fingerprint sizes (easy to get OOM errors here) and the combinatorics of different things to try, so I changed some inputs to give smaller fingerprint vectors and tested all combinations of R^2 and RMSE metrics, RF vs fully-connected network models, and Voxelized ECFP vs Ligand ECFP fingerprints.
* Unsurprisingly, optimized AtomicConvs outperform all of the above combinations. Optimizing the ECFP settings could almost certainly boost performance further, but I stuck with default settings.

CCing @rbharath @mufeili

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,6 +32,14 @@
...
+
+| Rank | Model         | Featurization  | Test RMSE (R^2)     | Validation RMSE (R^2) | Contact                           | References                                                                               | Date           |
+| ---- | ------------- | -------------- | ---------------- | ------------------ | --------------------------------- | ---------------------------------------------------------------------------------------- | -------------- |
+| 1    | ACNN | AtomicConv | 1.8604 +- 0.0540 (0.5405 +- 0.0124) | 1.5885 +- 0.0865 (0.6015 +- 0.0330)   | [Nathan Frey](n.frey@seas.upenn.edu) | [Paper](https://arxiv.org/pdf/1703.10603.pdf), [Code](./examples) | March 24, 2021  |
+| 2    | RF | Voxelized ECFP | 2.2488 +- 0.0615 (0.0773 +- 0.0242) | 2.1411 +- 0.0081 (0.0176 +- 0.0084)   | [Nathan Frey](n.frey@seas.upenn.edu) | [Paper](https://pubs.acs.org/doi/pdf/10.1021/ci100050t), [Code](./examples) | March 31, 2021  |
+| 3    | MTR | Ligand ECFP | 2.2596 +- 0.5495 (0.0440 +- 0.0042) | 2.5236 +- 0.2842 (0.1947 +- 0.0004)   | [Nathan Frey](n.frey@seas.upenn.edu) | [Paper](https://pubs.acs.org/doi/pdf/10.1021/ci100050t), [Code](./examples) | March 31, 2021  |
```
Comment (mufeili) [inline README.md:41]: I wonder if MLP is more appropriate. Any multitask regression model is an MTR, including ACNN.

Comment (mufeili) [inline README.md:41]: By "Ligand ECFP", can I assume that the pocket information is not considered?

File: examples/README.md [hunk 2]
```diff
@@ -3,6 +3,7 @@
 The repository includes the following example scripts:
```
Comment (mufeili) [inline examples/README.md:3]: I wonder if it will be better to have two separate folders, one for single-molecule property prediction, one for pdbbind.

File: examples/README.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
...
 - fingerprint
 - gnn
+- acnn
```
Comment (mufeili) [inline examples/README.md:6]: Do we need to add "Interaction Fingerprint"?

File: examples/utils.py [hunk 4]
```diff
@@ -62,6 +62,18 @@ def load_dataset(args):
...
     from deepchem.feat import MolGraphConvFeaturizer
     featurizer = MolGraphConvFeaturizer()
+  elif args['featurizer'] == 'AC':
+    from deepchem.feat import AtomicConvFeaturizer
+    featurizer = AtomicConvFeaturizer(
+        frag1_num_atoms=100,
+        frag2_num_atoms=1000,
+        complex_num_atoms=1100,
+        max_num_neighbors=12,
+        neighbor_cutoff=4)
```
Comment (mufeili) [inline examples/utils.py:72]: I wonder if we should treat `max_num_neighbors` and `neighbor_cutoff` as hyperparameters as well.

File: examples/utils.py [hunk 5]
```diff
@@ -95,6 +107,16 @@ def load_dataset(args):
     from deepchem.molnet import load_lipo
     tasks, all_dataset, transformers = load_lipo(
         featurizer=featurizer, splitter=splitter, reload=False)
+  elif args['dataset'] == 'PDBbind':
+    from deepchem.molnet import load_pdbbind
+    tasks, all_dataset, transformers = load_pdbbind(
+        featurizer=featurizer,
+        save_dir='.',
+        data_dir='.',
+        splitter='random',
```
Comment (mufeili) [inline examples/utils.py:116]: Any deterministic split for PDBbind?

File: examples/configures/ACNN_AC/PDBbind_R2.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (mufeili) [inline examples/configures/ACNN_AC/PDBbind_R2.json:1]: Is this file used anywhere?

File: examples/README.md [hunk 7]
```diff
@@ -56,6 +57,46 @@ The feasible arguments include:
...
+- **Dataset**: `-d dataset`
+    - Specifies the dataset to use, which can be one of the following:
+        - `PDBbind`
+- **Hyperparameter Search (optional)**: `-hs`
+    - Perform a hyperparameter search using Bayesian optimization. It determines the best 
+      hyperparameters based on the validation metric averaged across 3 runs.
+    - If not specified, the script uses default hyperparameters.
+- **Number of Hyperparameter Search Trials (optional)**: `-nt num_trials` [default=16]
+    - The number of trials for hyperparameter search. This comes into effect only when the user specifies 
+      `-hs` as described above.
```
Comment (mufeili) [inline examples/README.md:78]: Add the doc for the `metric` argument.

File: examples/acnn.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+  args = parser.parse_args().__dict__
+
+  mkdir_p(args['result_path'])
+  if args['hyper_search']:
+    print('Start hyperparameter search with Bayesian '
+          'optimization for {:d} trials'.format(args['num_trials']))
+    val_metrics, test_metrics = bayesian_optimization(args)
+  else:
+    print('Use the manually specified hyperparameters')
+    with open('configures/{}_{}/{}.json'.format(
```
Comment (mufeili) [inline examples/acnn.py:247]: This file should also consider the metric to use.

File: examples/README.md [hunk 9]
```diff
@@ -56,6 +57,46 @@ The feasible arguments include:
...
```
Comment (mufeili) [inline examples/README.md:98]: Add the doc for the model, the featurizer and the metric.

Comment (mufeili) [inline examples/README.md:98]: Add the doc for the model, the featurizer, and the metric.

File: examples/interaction_fingerprint.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+
+  mkdir_p(args['result_path'])
+  if args['hyper_search']:
+    print('Start hyperparameter search with Bayesian '
+          'optimization for {:d} trials'.format(args['num_trials']))
+    val_metrics, test_metrics = bayesian_optimization(args)
+  else:
+    print('Use the manually specified hyperparameters')
+    with open('configures/{}_{}/{}.json'.format(
+        args['model'], args['featurizer'], args['dataset'])) as f:
```
Comment (mufeili) [inline examples/interaction_fingerprint.py:276]: need to consider metric?

File: examples/configures/MTR_ECFP/PDBbind.json [hunk 11]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (mufeili) [inline examples/configures/MTR_ECFP/PDBbind.json:1]: Where did you use this?

File: examples/configures/RF_ECFP/PDBbind.json [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (mufeili) [inline examples/configures/RF_ECFP/PDBbind.json:1]: Ditto

File: examples/interaction_fingerprint.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+      'Options include 1) random forest (RF), 2) multitask regressor (MTR) (default: RF)'
+  )
+  parser.add_argument(
+      '-f',
+      '--featurizer',
+      choices=['flat_combined', 'voxel_combined', 'all_combined',
+      'ecfp_ligand', 'ecfp_hashed', 'ecfp', 'splif'],
+      default='flat_combined',
+      help=
+      'Options include 1) flat (ecfp, splif, hbond), 2) voxel (ecfp, splif, salt bridge, charge, hbond, pi stack, cation pi), 3) all (default: flat)'
```
Comment (mufeili) [inline examples/interaction_fingerprint.py:224]: The options in the doc seem to be different from the options in `choices`.

File: examples/interaction_fingerprint.py [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+      '-r',
+      '--num-runs',
+      type=int,
+      default=3,
+      help='Number of runs for each hyperparameter configuration (default: 3)')
+  parser.add_argument(
+      '-pa',
+      '--patience',
+      type=int,
+      default=10,
```
Comment (mufeili) [inline examples/interaction_fingerprint.py:242]: Why did you use 10 instead? Also the doc remains to be the same.

File: examples/interaction_fingerprint.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+from copy import deepcopy
+from functools import partial
+from hyperopt import hp, fmin, tpe
+from shutil import copyfile
+from sklearn.ensemble import RandomForestRegressor
+
+from utils import init_trial_path, load_dataset, EarlyStopper
+
+
+def rf_model_builder(model_dir, hyperparams, mode):
```
Comment (mufeili) [inline examples/interaction_fingerprint.py:14]: This function is redundant now.

File: examples/interaction_fingerprint.py [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+
+def load_model(save_path, n_features, args, tasks, hyperparams):
+  if args['dataset'] in ['PDBbind']:
+    mode = 'regression'
+  else:
+    raise ValueError('Unexpected dataset: {}'.format(args['dataset']))
+
+  if args['model'] == 'RF':
+    # model = dc.models.SingletaskToMultitask(
+    #     tasks, partial(rf_model_builder, hyperparams=hyperparams, mode=mode))
```
Comment (mufeili) [inline examples/interaction_fingerprint.py:32]: You can remove L31-32.

File: examples/interaction_fingerprint.py [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+    model = load_model(save_path, n_features, args, tasks, hyperparams)
+
+    if args['model'] == 'RF':
+      model.fit(train_set)
+    else:
+      # Object for early stop tracking
+      stopper = EarlyStopper(save_path, args['metric'], args['patience'])
+
+      # 50 for maximum number of epochs
+      for epoch in range(50):
```
Comment (mufeili) [inline examples/interaction_fingerprint.py:84]: Why 50 rather than 100?

File: README.md (hunk 1 above)
Comment (ncfrey) [reply README.md:41]: Changed to MLP

Comment (ncfrey) [reply README.md:41]: Yes, that's correct.

File: examples/README.md (hunk 2 above)
Comment (ncfrey) [reply examples/README.md:3]: Sounds good - I will reorganize into folders

File: examples/README.md (hunk 3 above)
Comment (ncfrey) [reply examples/README.md:6]: Added

File: examples/utils.py (hunk 4 above)
Comment (ncfrey) [reply examples/utils.py:72]: I think that would be best, but it would involve re-featurizing pdbbind for every iteration of model training, which would be pretty time-consuming. These values are taken from the original ACNN paper.

File: examples/utils.py (hunk 5 above)
Comment (ncfrey) [reply examples/utils.py:116]: Not yet, but there is a plan to work on splits based on interaction fingerprints and ligand fingerprints!

File: examples/configures/ACNN_AC/PDBbind_R2.json (hunk 6 above)
Comment (ncfrey) [reply examples/configures/ACNN_AC/PDBbind_R2.json:1]: It isn't currently used, but it could be used to re-run the benchmark with optimal hyperparams.

File: examples/README.md (hunk 7 above)
Comment (ncfrey) [reply examples/README.md:78]: Added

File: examples/acnn.py (hunk 8 above)
Comment (ncfrey) [reply examples/acnn.py:247]: Fixed

File: examples/README.md (hunk 9 above)
Comment (ncfrey) [reply examples/README.md:98]: Added

File: examples/interaction_fingerprint.py (hunk 10 above)
Comment (ncfrey) [reply examples/interaction_fingerprint.py:276]: Yes, added

File: examples/configures/MTR_ECFP/PDBbind.json (hunk 11 above)
Comment (ncfrey) [reply examples/configures/MTR_ECFP/PDBbind.json:1]: I put these configuration files in so that the benchmarks can be re-run without hyperparam optimization. I can remove if needed.

File: examples/interaction_fingerprint.py (hunk 13 above)
Comment (ncfrey) [reply examples/interaction_fingerprint.py:224]: Updated!

File: examples/interaction_fingerprint.py (hunk 14 above)
Comment (ncfrey) [reply examples/interaction_fingerprint.py:242]: For pdbbind I found that the validation error converges rapidly and the models overfit quickly. It also saves a good amount of time to decrease the patience, because each epoch takes awhile. I'll update the doc with this default value.

File: examples/interaction_fingerprint.py (hunk 15 above)
Comment (ncfrey) [reply examples/interaction_fingerprint.py:14]: Removed

File: examples/interaction_fingerprint.py (hunk 16 above)
Comment (ncfrey) [reply examples/interaction_fingerprint.py:32]: Removed

File: examples/interaction_fingerprint.py (hunk 17 above)
Comment (ncfrey) [reply examples/interaction_fingerprint.py:84]: Same as above - these models overfit to pdbbind very quickly and in practice I've never found that > 20 epochs are needed. Also the training is pretty time consuming when doing hyperparam optimization.

**Comment by ncfrey** [comment]
Thanks @mufeili for the detailed review! I have addressed the comments above.

CCing @rbharath

**Review by rbharath — APPROVED** [review]
This looks great! If @mufeili is done with his review, we can go ahead and merge this in

**Comment by ncfrey** [comment]
Closed by #34
